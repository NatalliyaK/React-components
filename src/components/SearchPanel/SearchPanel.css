.search {
    display: flex;
    justify-content: center;
}

.search-control {
    width: 45%;
    min-height: 35px;
    background: #86c2b5;
    border: solid 2px gray;
    padding: 8px;
    margin: 20px;
}

.card-list {
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
    flex-direction: column;
    padding: 10px;
}

.card-list__button {
    width: 10%;
}