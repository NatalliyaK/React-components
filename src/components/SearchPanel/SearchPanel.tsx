import React, { Component } from 'react';
import StarWarsService from '../component/StarWarsService/StarWarsService';
import CardList from '../CardList';
import Modal from '../Modal/Modal';
import './SearchPanel.css';
import Spiner from '../Spiner/Spiner';

type TsearchPanelState = {
  value: string;
  isLoaded: boolean;
  error: boolean;
  showModal: boolean;
  selectedNumber: number;
  data: {
    name: string;
    birth_year: string;
    eye_color: string;
    height: string;
    mass: string;
    hair_color: string;
    gender: string;
    skin_color: string;
  }[];
};

type TsearchPanelProps = { [key: string]: unknown };

export default class SearchPanel extends Component<TsearchPanelProps, TsearchPanelState> {
  constructor(props: TsearchPanelProps) {
    super(props);
    this.state = {
      value: '',
      isLoaded: false,
      error: false,
      data: [],
      showModal: false,
      selectedNumber: 0,
    };
  }

  handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
  };

  StarWarsService = new StarWarsService();

  componentDidMount() {
    const searchQuery = localStorage.getItem('searchQuery') ?? '';

    if (searchQuery) {
      this.setState({ value: searchQuery });
    }
  }

  componentWillUnmount() {
    localStorage.setItem('searchQuery', this.state.value);
  }

  handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const searchValue = event.target.value;

    this.StarWarsService.getSearchResults(searchValue).then((response) => {
      this.setState({ data: response.results });
    });
    this.setState({ value: event.target.value });
  };

  handleModal = (event: React.MouseEvent<HTMLButtonElement>): void => {
    const clickedButton = event.currentTarget;
    this.setState({ showModal: !this.state.showModal });
    this.setState({
      selectedNumber: parseInt(
        typeof clickedButton.dataset.cardNumber === 'string' ? clickedButton.dataset.cardNumber : ''
      ),
    });
  };

  render() {
    if (this.state.isLoaded) {
      <Spiner />;
    } else {
      return (
        <>
          <form className="search" onSubmit={this.handleSubmit}>
            <input
              className="search-control"
              type="text"
              placeholder="Search"
              onChange={this.handleChange}
              value={this.state.value}
            />
          </form>
          {Object.keys(this.state.data).map((item, i) => (
            <div className="card-list" key={i}>
              <CardList cardData={this.state.data[i]} />
              <button
                className="card-list__button"
                onClick={this.handleModal}
                data-card-number={item}
              >
                {' '}
                Show more{' '}
              </button>
            </div>
          ))}
          {this.state.showModal && (
            <Modal
              onClick={this.handleModal}
              cardData={this.state.data[this.state.selectedNumber]}
            />
          )}
        </>
      );
    }
  }
}
