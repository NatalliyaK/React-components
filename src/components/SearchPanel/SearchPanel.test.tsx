import React from 'react';
import { render, screen } from '@testing-library/react';
import SearchPanel from './SearchPanel';

jest.spyOn(window.localStorage.__proto__, 'getItem');
jest.spyOn(window.localStorage.__proto__, 'setItem');

describe('SearchPanel component', () => {
  it('render correctly', () => {
    const { asFragment } = render(<SearchPanel />);

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByPlaceholderText(/Search/i)).toBeInTheDocument();
  });

  it('ls', () => {
    const { unmount } = render(<SearchPanel />);

    expect(localStorage.getItem).toBeCalledTimes(1);
    expect(localStorage.setItem).not.toBeCalled();

    unmount();
    expect(localStorage.setItem).toBeCalledTimes(1);
  });
});

describe('SearchPanel component', () => {
  test('it renders', () => {
    render(<SearchPanel />);

    expect(screen.getByText('Show more')).toBeInTheDocument();
  });
});
