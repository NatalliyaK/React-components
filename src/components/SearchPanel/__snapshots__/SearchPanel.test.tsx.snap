// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPanel component render correctly 1`] = `
<DocumentFragment>
  <form
    class="search"
  >
    <input
      class="search-control"
      placeholder="Search"
      type="text"
      value=""
    />
  </form>
</DocumentFragment>
`;
